dapr_agents/agents/agent/agent.py

```diff
@@ -1,235 +1,279 @@
 
         # Always print the last user message for context, even if no input_data is provided
-        if user_message:
-            self.text_formatter.print_message(user_message)
+        if user_message_copy is not None:
+            # Ensure keys are str for mypy
+            self.text_formatter.print_message(
+                {str(k): v for k, v in user_message_copy.items()}
+            )
 
-        # Process conversation iterations
+        # Process conversation iterations and return the result
         return await self.process_iterations(messages)
 
-    async def process_response(self, tool_calls: List[ToolCall]) -> None:
+    async def execute_tools(self, tool_calls: List[ToolCall]) -> List[ToolMessage]:
         """
-        Asynchronously executes tool calls and appends tool results to memory.
+        Executes a batch of tool calls in parallel, bounded by max_concurrent, using asyncio.gather (Python 3.10+ compatible).
+        Each tool call is executed asynchronously using run_tool, and results are appended to the persistent audit log (tool_history).
+        If any tool call fails, the error is propagated and other tasks continue unless you set return_exceptions=True.
 
         Args:
-            tool_calls (List[ToolCall]): Tool calls returned by the LLM.
+            tool_calls (List[ToolCall]): List of tool calls returned by the LLM to execute in this batch.
+            max_concurrent (int, optional): Maximum number of concurrent tool executions (default: 5).
+
+        Returns:
+            List[ToolMessage]: Results for this batch of tool calls, in the same order as input.
 
         Raises:
-            AgentError: If a tool execution fails.
+            AgentError: If any tool execution fails.
         """
-        for tool_call in tool_calls:
-            function_name = tool_call.function.name
-            tool_id = tool_call.id
-            function_args = (
-                tool_call.function.arguments_dict
-            )  # Use the property to get dict
-
-            if not function_name:
-                logger.error(f"Tool call missing function name: {tool_call}")
-                continue
+        # Limiting concurrency to avoid overwhelming downstream systems
+        max_concurrent = 10
+        semaphore = asyncio.Semaphore(max_concurrent)
+
+        async def run_and_record(tool_call: ToolCall) -> ToolMessage:
+            """
+            Executes a single tool call, respecting the concurrency limit.
+            Appends the result to the persistent audit log.
+            If the function name is missing, returns a ToolMessage with error status and raises AgentError.
+            """
+            async with semaphore:
+                function_name = tool_call.function.name
+                tool_id = tool_call.id
+                function_args = tool_call.function.arguments_dict
+
+                if not function_name:
+                    error_msg = f"Tool call missing function name: {tool_call}"
```

Comment:
Eventually we'll need to update this to support other types of tools - just mental note to self as reviewing